federation/pkg/federation-controller/service/endpoint_helper.go

```diff
@@ -0,0 +1,136 @@
+					//return err
+				}
+			}
+		}()
+	}
+}
+
+// Whenever there is change on endpoint, the federation service should be updated
+func (cc *clusterClientCache) syncEndpoint(key, clusterName string, clusterCache *clusterCache, serviceCache *serviceCache, fedClient federation_release_1_3.Interface) error {
+	cachedService, ok := serviceCache.get(key)
+	if !ok {
+		// here we filtered all non-federation services
+		return nil
+	}
+	// obj holds the latest service info from apiserver
+	obj, exists, err := clusterCache.endpointStore.GetByKey(key)
+	if err != nil {
+		glog.Infof("Unable to retrieve endpoint %v from store: %v", key, err)
+		clusterCache.endpointQueue.Add(key)
+		return err
+	}
+	if !exists {
+		// service absence in store means watcher caught the deletion, ensure LB info is cleaned
+		glog.Infof("endpoint has been deleted %v", key)
+		cc.processEndpointDeletion(cachedService, key, clusterName)
+	}
+	if exists {
+		endpoint, ok := obj.(*v1.Endpoints)
+		if ok {
+			glog.V(4).Infof("Found endpoint for federation service %s/%s from cluster %s", endpoint.Namespace, endpoint.Name, clusterName)
+			cc.processEndpointUpdate(cachedService, endpoint, fedClient, clusterName)
+		} else {
+			tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+			if !ok {
+				return fmt.Errorf("object contained wasn't a service or a deleted key: %+v", obj)
+			}
+			glog.Infof("Found tombstone for %v", key)
+			cc.processEndpointDeletion(cachedService, tombstone.Key, clusterName)
+		}
+	}
+
+	return nil
+}
+
+func (cc *clusterClientCache) processEndpointDeletion(cachedService *cachedService, obj interface{}, clusterName string) {
+	glog.V(4).Infof("Processing endpoint update for %s, cluster %s", obj, clusterName)
+	cachedService.rwlock.Lock()
+	defer cachedService.rwlock.Unlock()
+	_, ok := cachedService.endpointMap[clusterName]
+	if ok {
+		// endpoints lost, clean dns record
+		glog.V(4).Infof("cached endpoint was not found for %s, cluster %s, building one", obj, clusterName)
+		//TODO
+		ensureDNSRecords(clusterName, cachedService)
+		delete(cachedService.endpointMap, clusterName)
+	}
+}
+
+// Update dns info when endpoint update event received
+// We do not care about the endpoint info, what we need to make sure here is len(endpoints.subsets)>0
+func (cc *clusterClientCache) processEndpointUpdate(cachedService *cachedService, endpoint *v1.Endpoints, fedClient federation_release_1_3.Interface, clusterName string) {
+	glog.V(4).Infof("Processing endpoint update for %s/%s, cluster %s", endpoint.Namespace, endpoint.Name, clusterName)
+	cachedService.rwlock.Lock()
+	defer cachedService.rwlock.Unlock()
+	_, ok := cachedService.endpointMap[clusterName]
+	if !ok {
+		// first time get endpoints, update dns record
+		glog.V(4).Infof("cached endpoint was not found for %s/%s, cluster %s, building one", endpoint.Namespace, endpoint.Name, clusterName)
+		cachedService.endpointMap[clusterName] = len(endpoint.Subsets)
```

Comment:
I think that we want the number of ready addresses across all subsets here, not the number of subsets?

https://htmlpreview.github.io/?https://github.com/kubernetes/kubernetes/blob/HEAD/docs/api-reference/v1/definitions.html#_v1_endpointsubset